File: csharp/Reservation/Application/ReservationService.cs
using Reservation.Diagnostics;
using Reservation.Domain.Model;
using Reservation.Infrastructure;

namespace Reservation.Application
{
    public class ReservationService(ILogger logger, IMetrics metrics, InMemoryHotelRepository inMemoryHotelRepository)
    {
        public Result<ConfirmedReservation> Make(string hotelName, int paxNumber, (DateTime begin, DateTime end) stay)
        {
            logger.Debug($"{nameof(ReservationService)} : Start making reservation");

            var registeredHotel = inMemoryHotelRepository.IsRegistered(hotelName);
            if (registeredHotel == null)
            {
                var hotelNotRegisteredMessage = $"{nameof(ReservationService)} : Hotel {hotelName} is not registered in our reservation system";
                logger.Debug(hotelNotRegisteredMessage);
                return Result.Fail<ConfirmedReservation>(hotelNotRegisteredMessage);
            }

            if (!registeredHotel.HasAvailableRooms(stay))
            {
                var roomNotAvailableMessage = $"Hotel {hotelName} doesn't have room for the selected period";
                logger.Debug($"{ nameof(ReservationService)} : {roomNotAvailableMessage}");
                metrics.Increment($"{hotelName}.reservations.roomNotAvailable");
                return Result.Fail<ConfirmedReservation>(roomNotAvailableMessage);
            }

            logger.Debug($"{nameof(ReservationService)} : Hotel {hotelName} has available rooms");

            if (!registeredHotel.HasRoomForOccupancy(paxNumber))
            {
                var noRoomFoundMessage = $"No room found for {paxNumber} people";
                logger.Debug($"{nameof(ReservationService)} : {noRoomFoundMessage}");
                metrics.Increment($"{hotelName}.reservations.occupancyNotAvailable");
                return Result.Fail<ConfirmedReservation>(noRoomFoundMessage);
            }
            logger.Debug($"{nameof(ReservationService)} : Room found for {paxNumber} people");

            var reservation = registeredHotel.InitializeReservation(stay, paxNumber);
            logger.Debug($"{nameof(ReservationService)} : Room available from {stay.begin:yyyyMMdd} to {stay.end:yyyyMMdd}");

            var confirmedReservation = registeredHotel.Confirm(reservation);
            logger.Info($"{nameof(ReservationService)} : Reservation {confirmedReservation.Reference} is confirmed");

            metrics.Increment($"{hotelName}.reservations");

            return Result.Ok(confirmedReservation);
        }
    }
}

File: typescript-class/src/application/ReservationService.ts
import Logger from '../diagnostics/Logger'
import { Metrics } from '../diagnostics/Metrics'
import {
  HotelName,
  PaxNumber,
  RegisteredHotel,
  RegisteredHotelRoomAvailable,
  RegisteredHotelRoomOccupancyAvailable,
} from '../domain/hotel'
import { HotelRepository } from '../infrastructure/HotelRepository'

class ReservationService {
  private logger: Logger
  private metrics: Metrics
  private hotelRepository: HotelRepository

  constructor(
    logger: Logger,
    metrics: Metrics,
    hotelRepository: HotelRepository,
  ) {
    this.logger = logger
    this.metrics = metrics
    this.hotelRepository = hotelRepository
  }

  make(
    hotelName: HotelName,
    paxNumber: PaxNumber,
    stay: { begin: Date; end: Date },
  ) {
    this.logger.debug('Start making reservation')

    const registeredHotel = this.hotelRepository.isRegistered(hotelName)

    if (registeredHotel.type === 'unregisteredHotel') {
      const hotelNotFoundMessage = `Hotel ${registeredHotel.hotel} is not registered in our reservation system`
      this.logger.info(hotelNotFoundMessage)
      this.metrics.increment(`${hotelName}.reservations.hotelNotFound`)
      return { Error: hotelNotFoundMessage }
    }

    const hotelWithRoom = RegisteredHotel.hasAvailableRoom(
      registeredHotel,
      stay,
    )

    if (hotelWithRoom.type === 'registeredHotelRoomNotAvailable') {
      const roomNotAvailableMessage = `Hotel ${hotelWithRoom.hotel} doesn't have room for the selected period`
      this.logger.info(roomNotAvailableMessage)
      this.metrics.increment(
        `${hotelWithRoom.hotel}.reservations.roomNotAvailable`,
      )
      return { Error: roomNotAvailableMessage }
    }

    const hotelAvailability = RegisteredHotelRoomAvailable.hasRoomForOccupancy(
      hotelWithRoom,
      paxNumber,
    )

    this.logger.debug(`Hotel ${hotelWithRoom.hotel} has room available`)

    if (hotelAvailability.type === 'registeredHotelRoomOccupancyAvailable') {
      this.logger.debug(
        `Hotel ${hotelWithRoom.hotel}: Room found for ${paxNumber}`,
      )

      const draftReservation =
        RegisteredHotelRoomOccupancyAvailable.initializeReservation(
          hotelAvailability,
        )

      this.logger.debug(
        `Hotel ${hotelWithRoom.hotel}: Reservation option from ${stay.begin} to ${stay.end} for ${paxNumber} people`,
      )
      this.metrics.increment(`${hotelWithRoom.hotel}.reservations`)
      return RegisteredHotelRoomOccupancyAvailable.confirmReservation(
        hotelAvailability,
        draftReservation,
      )
    }

    const roomNotAvailableMessage = `No room found for ${paxNumber} people`
    this.logger.info(roomNotAvailableMessage)
    this.metrics.increment(
      `${hotelWithRoom.hotel}.reservations.occupancyNotAvailable`,
    )
    return { Error: roomNotAvailableMessage }
  }
}

export default ReservationService
